### ISSUE TITLE
[Proposal] Break Mapbox GL JS API Reference into multiple pages

### ISSUE THREAD

**Issue by katydecorah** [mapbox/mapbox-gl-js-docs#250]
The documentation team would like to propose breaking the [Mapbox GL JS API Reference](https://docs.mapbox.com/mapbox-gl-js/api/) into multiple pages. We recently completed this work with the [Style Specification](https://docs.mapbox.com/mapbox-gl-js/style-spec/) page.

By breaking the API Reference into multiple pages:

- 👍 Users will be able to find content easier through docs search and other search engines.
- 👍 We will be able to structure content to make it less overwhelming and more palatable to new users.
- 👍 Users will experience faster page loads.
- 👍 The sidebar will be easier to digest and navigate for users.
    - The sidebar will use [dr-ui](https://github.com/mapbox/dr-ui)’s SectionedNavigation component which is used by all documentation pages to help deliver a consistent experience for our users. 
    - We recently [rolled out improvements to this component](https://github.com/mapbox/mapbox-gl-js-docs/pull/245) so that the sidebar will scroll to the active item on load, which helps keep the active item in view for the user so they don’t lose their spot. You can see it in action on the [Style Specification](https://docs.mapbox.com/mapbox-gl-js/style-spec/) page.
- 👍 Our team will have improved data around usage. 
    - Each week the API Reference receives approximately 50,000 page views. Due to the page’s length, it’s difficult for us to determine what content is considered popular with users.
    - After breaking the Style Specification into pages we learned the the Layers page is the most popular Style Spec page.
- 👎 Users can no longer command+F on the entire API Reference since the content will be organized across pages. However, since the page has grown so long, we’ve had reports that command+F can fail to function on this page.

## Mock ups

The follow screenshots are examples of what the proposed pages and sidebar hierarchy could look like.

> ![image](https://user-images.githubusercontent.com/2180540/77317643-f578d680-6ce1-11ea-91f4-090bfe9708fb.png)
> Example of the Home page of API Reference

> ![image](https://user-images.githubusercontent.com/2180540/77317766-2fe27380-6ce2-11ea-8a67-ac05e5965459.png)
> Example of the Map page with all its properties organized in the sidebar. When a sidebar item is scrolled to or clicked, the item’s subordinate hierarchy appears. This interaction will be the same as how a page like [Layers](https://docs.mapbox.com/mapbox-gl-js/style-spec/layers/) reveals subordinate hierarchy on the same page.

> ![image](https://user-images.githubusercontent.com/2180540/77317812-4983bb00-6ce2-11ea-8b23-115e72f3c39c.png)
>Example of the Markers and controls page

## Proposed pages and content

The API Reference will be split into 8 pages:

1. **API Reference** - Introduction to the API with an overview or get started section to help orient new users.
2. **Map** - All `Map` methods, events, properties, utilities, and options. 
    - Since there are over 100 methods and events, we are currently exploring categories to help organize them all, such as: Get map state, Modify map state, Work with styles. We are continuing to refine these categories to make them meaningful to users and welcome suggestions.
    - Under each category, methods and events will be combined, sorted alphabetically, and assigned an icon to help differentiate between method and event, similar to how we’re handling the [Style Spec Layers page](https://docs.mapbox.com/mapbox-gl-js/style-spec/layers/#symbol). 
3. **Properties and options** - Tentatively named, this section includes all `mapboxgl` namespaced properties, utilities, and options.
4. **Markers and controls** - This page rebrands the “User Interface” to help satisfy just add a maker to a map (JAMTAM).
5. **Geography and geometry** - No changes from the original section.
6. **User interaction handlers** - No changes from the original section.
7. **Sources** - No changes from the original section.
8. **Events** - No changes from the original section.

## Timeline

The documentation team is hoping to complete this work before the upcoming examples sprint in April, but are flexible around this date.

## Feedback

We look forward to your feedback, ideally by Friday (3/20): @mapbox/gl-js @malwoodsantoro @andrewharvey @stevage 

cc @mapbox/docs

**Comment by asheemmamoowala** [comment]
@katydecorah In this proposed sidebar organization, it looks like properties and utilities that are part of the `mapboxgl` namespace are part of the `Map` class. Can those be separated out of `Map`. It would make it easier for readers to find and correctly use them.

![image](https://user-images.githubusercontent.com/11619675/76800898-cdddb780-6791-11ea-8531-d35683fb1b55.png)

**Comment by stevage** [comment]
Thanks very much for seeking feedback.

For me (as a power user?), the loss of intra-page searching is a bit of a bummer. But then, in the proposed breakdown, I think I'd be searching within the Map object maybe 80% of the time.  I rarely use any of the Sources, Handlers  or Geography/Geometry sections. (Although, hey, I just discovered that there's a lngLat.distanceTo() method which I wish I'd known about :))

So I think the negative impact is actually going to be fairly low.

> Since there are over 100 methods and events, we are currently exploring categories to help organize them all, such as: Get map state, Modify map state, Work with styles. We are continuing to refine these categories to make them meaningful to users and welcome suggestions.

Excellent. I think I have suggested this in the past.

> Under each category, methods and events will be combined, sorted alphabetically, and assigned an icon to help differentiate between method and event, similar to how we’re handling the Style Spec Layers page.

Hooray, thank god for alphabetic sorting. 

It looks like the biggest impact to me will be that the URL I visit all the time will now have a `/map` at the end of it. :)

**Comment by andrewharvey** [comment]
I'm also a fan of Ctrl+F searching the whole page, I use that quite a bit. I know there is a site wide search but sometimes I just want a light weight filter just within the GL JS API doc.

I find loading the whole page is faster because once the page is loaded you can navigate around quickly. I noticed on the new style spec page, there is a lag when opening sections. Perhaps these could be preloaded?

Also on the new style spec page, I'd prefer those expansion panels on the left to be expanded by default so I can quickly skim through without needing to expand each one (there is a one to two second lag just to open it), so same would apply to the API doc if they moved to the same style layout.

Either way, these aren't blockers to moving to multiple pages.

**Comment by stevage** [comment]
>I know there is a site wide search but sometimes I just want a light weight filter just within the GL JS API doc.

Yeah. FWIW, I haven't been finding the sitewide search very effective. (Not a criticism - search is ridiculously hard to do well). A search just within the API and style spec page would be amazing.

**Comment by andrewharvey** [comment]
After observing a bit more on how I'm using the docs, I will quite frequently want to check something, eg. the syntax of `addLayer` so will go to the api docs page, then Ctrl+F `addLayer`. So if the page will be broken up, it would be really nice to have a filter search, activated by `/` then searches on keywords like addlayer to take me straight there.

**Comment by stevage** [comment]
My understanding is that that would still work as now: you'll be able to use Ctrl+F for anything within the Map object, which is most of the API really.

**Comment by katydecorah** [comment]
> In this proposed sidebar organization, it looks like properties and utilities that are part of the mapboxgl namespace are part of the Map class. Can those be separated out of Map. It would make it easier for readers to find and correctly use them.

@asheemmamoowala You're right, these `mapboxgl` properties and utilities don't really fit on the Map page. It seems this content is enough for its own page. I'm going to tentatively add another page to split this content. (I'll update the original post to reflect this change).

> > Under each category, methods and events will be combined, sorted alphabetically, and assigned an icon to help differentiate between method and event, similar to how we’re handling the Style Spec Layers page.

> Excellent. I think I have suggested this in the past.

@stevage You did! Thank you for your helpful feedback and ideas.

@andrewharvey This is all great feedback. Right now we don't have the capacity to add a more advanced filter during this iteration. However, we have received this feedback before and the docs team will give it more consideration. On the subject of page speed, this is also something the docs team has started thinking about and we are hoping to make more improvements to Batfish, the static site generator, this year.

**Comment by danswick** [comment]
While working on some inline code snippets, I was thinking that we might want to consider changing the way we document events. 

Users can end up in a section on a particular event type and be unsure what that event pertains to or how to set up a listener for it. 

It might be nice to create an events component that supported both a `membersList`-style list of events as well as a longer-form intro section with a long-ish example showing how to use those events. This is a bit of a challenge now since there isn't really a way to control each section in such a fine-grained way.

**Comment by katydecorah** [comment]
We’re ready for @mapbox/gl-js @mapbox/docs @andrewharvey @stevage to review our proposed changes to split the API reference into multiple pages.

- **How do I view the changes?** For Mapbox team, visit the staging site or you can build https://github.com/mapbox/mapbox-gl-js-docs/pull/272 locally (`nvm use && yarn && yarn start`).
- **How do I submit feedback?** Comment on this issue or leave code comments on https://github.com/mapbox/mapbox-gl-js-docs/pull/272.
- **What's the deadline to submit feedback?** End of the week 🙏 


## Changes and additions to original proposal


**Removed the “Overview” page and combined its content into the API Reference’s introduction page.** Since much of the content was already duplicated, this helps provide a single location to onboard users to Mapbox GL JS.

> ![image](https://user-images.githubusercontent.com/2180540/82345477-4e53ab00-99c3-11ea-956f-0c0f5213934b.png)




**Added a search to all instance or events members with > 9 items.** We originally planned to organize Map’s members into categories, but as we were developing, we didn’t find that this solution would improve the user’s ability to find information. The search uses the member’s title and description to return results. We also decided against adding every single member item to the sidebar, since it's overwhelming and makes the sidebar difficult to navigate.

> ![image](https://user-images.githubusercontent.com/2180540/82344980-b5bd2b00-99c2-11ea-9098-bd895518dd00.png)



**Improved legibility of members.** We made some style improvements to make the contents of each member easier to read.

Status | Image
---|---
Current (before) | ![image](https://user-images.githubusercontent.com/2180540/82345039-c9689180-99c2-11ea-9383-7445e335ecbe.png)
Proposed (after) | ![image](https://user-images.githubusercontent.com/2180540/82345022-c53c7400-99c2-11ea-9992-ef47a56f8dda.png)
 
**Removed originally planned group of mapboxgl namespaced items**. We had planned to organized the Properties and options page into section, but found that the groups added little context and were hard to uncover in the sidebar. All mapboxgl namespaced items appear in the sidebar.

> ![image](https://user-images.githubusercontent.com/2180540/82346284-40eaf080-99c4-11ea-96bc-d6e092bfaa95.png)

**Comment by stevage** [comment]
Where is "the staging site"?

**Comment by andrewharvey** [comment]
> I take it there is no public-facing staging site where changes like this can be reviewed?

It says there "For Mapbox team" so you and I will need to build and run locally to test.

**Comment by andrewharvey** [comment]
I can only really comment from my power user point of view, but my observations are below. I'm not trying to be tough, but just my honest thoughts when taking it for a test drive.

- A filter on the left panel exactly like the Examples and Plugins pages, but for API reference and Style Spec pages would be very handy. I know I mentioned this earlier and noted that it's not targeted for this update, but it's probably the top feature in my mind for a power user. I use Ctrl+F all the time on the current docs, eg I might search for Popup to go straight to the Popup class.

- Maybe it's just me but after opening an accordion like this, the styling makes it look like this is part of the Introduction accordion, but actually it's part of the Map accordion.

![Screenshot from 2020-05-21 18-11-06](https://user-images.githubusercontent.com/117278/82538084-865b1980-9b8e-11ea-8333-8ca253c6d4b3.png)

- I feel like having all the parameters, methods and events listed in the left panel would make it easier for me to quickly scan and find the one I want.

![Screenshot from 2020-05-21 18-13-40](https://user-images.githubusercontent.com/117278/82538573-49dbed80-9b8f-11ea-8adc-4d1d93560031.png)

- I'd probably prefer having all these boxes expanded by default, it would make it easier to quickly scan through.

- I noticed some of the link don't work, eg. under Map on the left clicking example the link is to `#map-example` but in the document the actual anchor is `map-class-example`

![Screenshot from 2020-05-21 18-17-02](https://user-images.githubusercontent.com/117278/82538597-552f1900-9b8f-11ea-8d04-5e698b6ed2f1.png)

- Also mentioned before, but I'm seeing a 1.25sec delay due to the network request when opening an according, although not the end of the world, it is a bit annoying, so anything to preload these chunks in the background after the first screen/page is done would improve the usability.

- Merging the Overview and API pages, makes sense to me. Before the list of pages was a bit long, and now you naturally flow straight into it after the intro.

**Comment by andrewharvey** [comment]
Also to add, as a power user most of the time I open the docs is to search for something specific, just now I wanted to check the documentation for cluster and the options like clusterRadius. Right now I have to know that

1. It's on the style spec page
2. It's under sources
3. Its under geojson

While I guess search is ideal for this as I can search for "cluster radius", a few things turned be off the site search

1. by default it does a global search, and so pulls up android sdk mixed in, I know I can filter that down to just GL JS but search feels too heavyweight sometimes.
2. individual properties like clusterRadius don't show up in the search.

So most of the time I know roughly what the name of the term is and just want to quickly go straight to the documentation for it.

**Comment by stevage** [comment]
I'm not sure I will find the effort to set this up locally before the deadline, but I'm pretty sure I'm going to agree with everything Andrew says.

In a nutshell for me:

- being able to search is everything
- currently the search as implemented doesn't really meet my needs, because it includes a lot of stuff that will never be relevant.

Probably the easiest solution is if a search box within mapbox-gl-js docs defaults to only searching within mapbox-gl-js.

One of the things I really like about the [OpenLayers documentation](https://openlayers.org/en/latest/apidoc/index.html) is the filter box in the top left. Type in a word like "icon" and it quickly finds classes, methods, properties that contain the word. It's incredibly efficient to use.

**Comment by katydecorah** [comment]
@andrewharvey @stevage thank you both for your feedback. We're continuing to work through them, but wanted to note that we just added an API reference search to the sidebar #326 

![image](https://user-images.githubusercontent.com/2180540/82708347-05faee00-9c4c-11ea-9722-32cada44a117.png)

We're hoping that it will help bridge the gap between power users and new users. If this new search is successful, we will investigate adding something similar to the Style Specification sidebar.

**Comment by andrewharvey** [comment]
@katydecorah The new sidebar filter looks awesome glad to see it made the cut.

**Comment by stevage** [comment]
Yeah, that looks really great.

FWIW I had a quick attempt to build this, but something failed and I'd rather not go down that rabbit hole. Would be great to have a public staging site for future review requests.

**Comment by asheemmamoowala** [comment]
@katydecorah Thanks for putting together this ticket to collect feedback. The break-up of the API reference into the new sections feels great. Once this lands, we'll want to add more introduction/context to each of the sections. For ex when clicking into the `Markers and Controls` section - I'm unsure of what I should be looking for - and whether to navigate within the content panel or go back to the left panel. 

![image](https://user-images.githubusercontent.com/11619675/83074579-4204c900-a027-11ea-9be2-a13850ccf542.png)


### Left Panel

> I feel like having all the parameters, methods and events listed in the left panel would make it easier for me to quickly scan and find the one I want.

@andrewharvey I think this is definitely true for  the `mapboxgl` namespace, `Map` and `Marker`. What I find hard to see the benefit of is the expansion of the method/property in the left panel to link to its Parameters, Example, Related sections (see image below) - because I have to click on the item to reveal those, which also navigates to all that section.

![image](https://user-images.githubusercontent.com/11619675/83074344-d589ca00-a026-11ea-8727-b29e759071a3.png)


@katydecorah The left panel makes it hard to identify the context for the Properties and Options section. At first glance, it looks like it reveals the properties of the `Map` object, but one has to look at a code example to know that these are on the `mapboxgl` namespace. 

![image](https://user-images.githubusercontent.com/11619675/83076516-8e9dd380-a02a-11ea-8345-434de632351d.png)

### Questions
- Can the `Map` instance members be surfaced in the left panel - either by hoisting into their own section or expanding in place `Instance members`? Maybe separating the page for Map constructor/options from the instance members would be a side benefit of this.
- Would you consider using `Globals` instead of `Properties and Options` to better capture what the items in that section reprensent?

**Comment by katydecorah** [comment]
Thank you @stevage @andrewharvey @asheemmamoowala for your feedback.

> A filter on the left panel exactly like the Examples and Plugins pages, but for API reference and Style Spec pages would be very handy. I know I mentioned this earlier and noted that it's not targeted for this update, but it's probably the top feature in my mind for a power user. I use Ctrl+F all the time on the current docs, eg I might search for Popup to go straight to the Popup class.

> currently the search as implemented doesn't really meet my needs, because it includes a lot of stuff that will never be relevant.

@andrewharvey  @stevage We added a search to the API Reference sidebar 🎉 

---

> Maybe it's just me but after opening an accordion like this, the styling makes it look like this is part of the Introduction accordion, but actually it's part of the Map accordion.

@andrewharvey I cannot unsee this now. So this is a standard component in our library, while redesigning it is out of scope for this project, I did create an issue to make design adjustments to make items look more visually distinct. https://github.com/mapbox/dr-ui/issues/280

---

> I feel like having all the parameters, methods and events listed in the left panel would make it easier for me to quickly scan and find the one I want.

> I think this is definitely true for the mapboxgl namespace, Map and Marker. What I find hard to see the benefit of is the expansion of the method/property in the left panel to link to its Parameters, Example, Related sections (see image below) - because I have to click on the item to reveal those, which also navigates to all that section.

@andrewharvey @asheemmamoowala  This will likely be a change for some power users.

While our original proposal was going to keep all items in the sidebar, we still felt the sidebar was too overwhelming and became unusable due to the long list of members. 

I also think that it's important to build a distinction between the function of the sidebar and the content page. If we have the sidebar list everything, then we are providing nearly duplicate views to the user in the sidebar and on the content page (take the current site as an example):

> ![image](https://user-images.githubusercontent.com/2180540/83185894-b27c1a80-a0f9-11ea-809b-7012a1262860.png)

The docs team sees the function of the sidebar to be a navigational tool to help you navigate the layout of the content page.


---

> I'd probably prefer having all these boxes expanded by default, it would make it easier to quickly scan through.

@andrewharvey We didn't have it in our proposal to make changes to the state of the item members and events. If you continue to feel strongly about this post-launch, would you mind opening an issue in this repo? Since we didn't receive any other feedback about the state of the boxes, we'd like to see if others feel similarly.

---

> I noticed some of the link don't work, eg. under Map on the left clicking example the link is to #map-example but in the document the actual anchor is map-class-example

@andrewharvey Thanks for the catch. This was fixed.

---

> Also mentioned before, but I'm seeing a 1.25sec delay due to the network request when opening an according, although not the end of the world, it is a bit annoying, so anything to preload these chunks in the background after the first screen/page is done would improve the usability.

@andrewharvey We'll be updating this repo with the latest Batfish version after this launch, we hope that will speed things up a little bit more. Our team is hoping to do more for performance improvements later this year.

---

> Can the Map instance members be surfaced in the left panel - either by hoisting into their own section or expanding in place Instance members? Maybe separating the page for Map constructor/options from the instance members would be a side benefit of this.

@asheemmamoowala It could be possible, but we think it would be confusing to split Map into two pages.

> Would you consider using Globals instead of Properties and Options to better capture what the items in that section represent?

@asheemmamoowala  We're hesitant as "Globals" is not as friendly to all users, especially those who are not full time developers. We chose "Properties and options" as a plain language alternative.







cc @danswick
